### PR TITLE
PKCE-enhanced OAuth2 code flow

### LINKED ISSUE #1656: Support PKCE extension in OAuth2 client
PKCE extension improves security of OAuth authentication flow and is frequent default for OAuth 2 servers in the wild, but Misago doesn't support it.

### PR DESCRIPTION
**Pull request by raydeal**
Added PKCE (Proof Key for Code Exchange) to authorization code flow of OAuth2
fix #1656

### CODE REVIEW

File: misago/oauth2/client.py [hunk 1]
```diff
@@ -25,6 +28,12 @@ def create_login_url(request):
         "state": state,
     }
 
+    if request.settings.oauth2_enable_pkce:
+        code_verifier = secrets.token_urlsafe()
+        request.session[SESSION_CODE_VERIFIER] = code_verifier
+        quote["code_challenge"] = get_code_challenge(code_verifier, request.settings.oauth2_pkce_code_challenge_method)
```
Comment (rafalp) [inline misago/oauth2/client.py:34]: Can you also rename `quote` to `querystring`?

File: misago/conf/migrations/0012_add_oauth2_pkce_settings.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    for setting in settings:
+        Setting.objects.filter(setting=setting["setting"]).delete()
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("misago_conf", "0011_add_notifications_settings"),
+    ]
+
+    operations = [migrations.RunPython(create_settings, remove_settings)]
```
Comment (rafalp) [inline misago/conf/migrations/0012_add_oauth2_pkce_settings.py:41]: Format code with black

**Comment by rafalp** [comment]
No contract tests to verify PKCE is included in requests to OAuth2 server?

File: misago/conf/admin/forms/oauth2.py [hunk 3]
```diff
@@ -203,6 +206,26 @@ class OAuth2SettingsForm(SettingsForm):
         max_length=200,
         required=False,
     )
+    oauth2_enable_pkce = YesNoSwitch(
+        label=pgettext_lazy("admin oauth2 settings form", "Enable OAuth2 PKCE"),
+        help_text=pgettext_lazy(
+            "admin oauth2 settings form",
+            "Enabling OAuth2 PKCE (Proof Key for Code Exchange) will add additional protection during code to token exchanges if OAuth2 server supports it.",
```
Comment (rafalp) [inline misago/conf/admin/forms/oauth2.py:213]:
```suggestion
            "Enabling this option will make Misago's OAuth2 client use PKCE (Proof Key for Code Exchange), increasing the security of the authentication process. The OAuth 2 server must also support PKCE.",
```

File: misago/conf/admin/forms/oauth2.py [hunk 4]
```diff
@@ -203,6 +206,26 @@ class OAuth2SettingsForm(SettingsForm):
...
+        ),
+        required=False,
+    )
+    oauth2_pkce_code_challenge_method = forms.ChoiceField(
+        label=pgettext_lazy("admin oauth2 settings form", "PKCE Code Challenge method"),
+        help_text=pgettext_lazy(
+            "admin oauth2 settings form",
+            "When PKCE is enabled this hashing method is used to generate the PKCE code challenge for OAuth2.",
```
Comment (rafalp) [inline misago/conf/admin/forms/oauth2.py:221]:
```suggestion
            "When PKCE is enabled this hashing method is used to generate the PKCE code challenge for the OAuth2 server.",
```

File: misago/conf/admin/forms/oauth2.py [hunk 5]
```diff
@@ -203,6 +206,26 @@ class OAuth2SettingsForm(SettingsForm):
...
+        ),
+        choices=[
+            ("S256", "S256"),
+            ("plain", "plain (no hashing)"),
```
Comment (rafalp) [inline misago/conf/admin/forms/oauth2.py:225]: "plain" choice should be marked for translation. You can use `admin oauth2 settings pkce choice` as a context.

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 6]
```diff
@@ -26,6 +26,8 @@ def test_oauth2_can_be_enabled(admin_client):
             "oauth2_json_name_path": "name",
             "oauth2_json_email_path": "email",
             "oauth2_json_avatar_path": "avatar",
+            "oauth2_enable_pkce": "0",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:29]: Please move pkce options in dict to where they are in order of fields.

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 7]
```diff
@@ -52,6 +54,8 @@ def test_oauth2_can_be_enabled(admin_client):
     assert settings["oauth2_json_name_path"] == "name"
     assert settings["oauth2_json_email_path"] == "email"
     assert settings["oauth2_json_avatar_path"] == "avatar"
+    assert settings["oauth2_enable_pkce"] is False
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:57]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 8]
```diff
@@ -77,6 +81,8 @@ def test_oauth2_can_be_enabled_without_avatar(admin_client):
             "oauth2_json_name_path": "name",
             "oauth2_json_email_path": "email",
             "oauth2_json_avatar_path": "",
+            "oauth2_enable_pkce": "0",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:84]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 9]
```diff
@@ -98,7 +104,7 @@ def test_oauth2_can_be_enabled_without_avatar(admin_client):
     assert settings["oauth2_user_token_location"] == "HEADER"
     assert settings["oauth2_user_token_name"] == "access_token"
     assert settings["oauth2_user_extra_headers"] == ""
-    assert settings["oauth2_send_welcome_email"] == False
+    assert settings["oauth2_send_welcome_email"] is False
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:107]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 10]
```diff
@@ -126,6 +132,8 @@ def test_oauth2_cant_be_enabled_with_some_value_missing(admin_client):
         "oauth2_json_name_path": "name",
         "oauth2_json_email_path": "email",
         "oauth2_json_avatar_path": "",
+        "oauth2_enable_pkce": "0",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:135]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 11]
```diff
@@ -136,6 +144,8 @@ def test_oauth2_cant_be_enabled_with_some_value_missing(admin_client):
         "oauth2_user_token_location",
         "oauth2_user_extra_headers",
         "oauth2_send_welcome_email",
+        "oauth2_enable_pkce",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:147]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 12]
```diff
@@ -215,6 +225,8 @@ def test_oauth2_scopes_are_normalized(admin_client):
             "oauth2_json_name_path": "name",
             "oauth2_json_email_path": "email",
             "oauth2_json_avatar_path": "",
+            "oauth2_enable_pkce": "0",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:228]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 13]
```diff
@@ -246,6 +258,8 @@ def test_oauth2_extra_token_headers_are_normalized(admin_client):
...
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:261]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 14]
```diff
@@ -307,6 +321,8 @@ def test_oauth2_extra_user_headers_are_normalized(admin_client):
...
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:324]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 15]
```diff
@@ -460,3 +476,58 @@ def test_oauth2_extra_headers_are_validated_to_be_unique(admin_client):
...
+            "oauth2_user_method": "GET",
+            "oauth2_user_token_location": "HEADER",
+            "oauth2_user_token_name": "access_token",
+            "oauth2_user_extra_headers": "",
+            "oauth2_send_welcome_email": "",
+            "oauth2_json_id_path": "id",
+            "oauth2_json_name_path": "name",
+            "oauth2_json_email_path": "email",
+            "oauth2_json_avatar_path": "avatar",
+            "oauth2_enable_pkce": "1",
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:504]: Ditto

File: misago/conf/admin/tests/test_change_oauth2_settings.py [hunk 16]
```diff
@@ -460,3 +476,58 @@ def test_oauth2_extra_headers_are_validated_to_be_unique(admin_client):
...
+    assert settings["oauth2_user_method"] == "GET"
+    assert settings["oauth2_user_token_location"] == "HEADER"
+    assert settings["oauth2_user_token_name"] == "access_token"
+    assert settings["oauth2_user_extra_headers"] == ""
+    assert settings["oauth2_send_welcome_email"] is False
+    assert settings["oauth2_json_id_path"] == "id"
+    assert settings["oauth2_json_name_path"] == "name"
+    assert settings["oauth2_json_email_path"] == "email"
+    assert settings["oauth2_json_avatar_path"] == "avatar"
+    assert settings["oauth2_enable_pkce"] is True
```
Comment (rafalp) [inline misago/conf/admin/tests/test_change_oauth2_settings.py:532]: Ditto

File: misago/conf/migrations/0012_add_oauth2_pkce_settings.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+                data["python_type"], data.pop("wet_value")
+            )
+
+        Setting.objects.create(**data)
+
+
+def remove_settings(apps, _):
+    Setting = apps.get_model("misago_conf", "Setting")
+    for setting in settings:
+        Setting.objects.filter(setting=setting["setting"]).delete()
```
Comment (rafalp) [inline misago/conf/migrations/0012_add_oauth2_pkce_settings.py:37]: `Setting.objects.delete(setting__in=[...])`

File: misago/oauth2/client.py [hunk 18]
```diff
@@ -11,21 +14,32 @@
...
     }
 
-    return "%s?%s" % (request.settings.oauth2_login_url, urlencode(quote))
+    if request.settings.oauth2_enable_pkce:
+        code_verifier = token_urlsafe()
+        request.session[SESSION_CODE_VERIFIER] = code_verifier
+        querystring["code_challenge"] = get_code_challenge(
+            code_verifier, request.settings.oauth2_pkce_code_challenge_method
+        )
+        querystring[
```
Comment (rafalp) [inline misago/oauth2/client.py:38]: Replace with:

```
querystring["code_challenge_method"] = (
    request.settings.oauth2_pkce_code_challenge_method
)
```

File: misago/oauth2/exceptions.py [hunk 19]
```diff
@@ -123,3 +123,19 @@ class OAuth2UserDataValidationError(OAuth2ProviderError):
...
+        "oauth2 error",
+        "The OAuth2 authorization flow is missing code verifier.",
+    )
+
+
+class OAuth2NotSupportedHashMethodError(OAuth2Error):
+    recoverable = False
+    message = pgettext_lazy(
+        "oauth2 error",
+        "The OAuth2 code challenge generator does not support this hash method.",
```
Comment (rafalp) [inline misago/oauth2/exceptions.py:140]:
```suggestion
        "The OAuth2 code challenge hash method is not supported.",
```

File: misago/oauth2/tests/test_get_access_token.py [hunk 20]
```diff
@@ -5,7 +5,7 @@
 
 from ...conf.test import override_dynamic_settings
 from .. import exceptions
-from ..client import REQUESTS_TIMEOUT, get_access_token
+from ..client import REQUESTS_TIMEOUT, get_access_token, SESSION_CODE_VERIFIER
```
Comment (rafalp) [inline misago/oauth2/tests/test_get_access_token.py:8]:
```suggestion
from ..client import (
    REQUESTS_TIMEOUT,
    SESSION_CODE_VERIFIER,
    get_access_token,
)
```

File: misago/oauth2/tests/test_login_url_creation.py [hunk 21]
```diff
@@ -1,8 +1,8 @@
-from unittest.mock import Mock
+from unittest.mock import Mock, patch
 from urllib.parse import parse_qsl, urlparse
 
 from ...conf.test import override_dynamic_settings
-from ..client import SESSION_STATE, create_login_url
+from ..client import SESSION_STATE, create_login_url, SESSION_CODE_VERIFIER
```
Comment (rafalp) [inline misago/oauth2/tests/test_login_url_creation.py:5]:
```suggestion
from ..client import (
    SESSION_CODE_VERIFIER,
    SESSION_STATE,
    create_login_url,
)
```

**Review by rafalp — CHANGES_REQUESTED** [review]
LGTM, I've left some cosmetics

File: misago/conf/migrations/0012_add_oauth2_pkce_settings.py (hunk 17 above)
Comment (raydeal) [reply misago/conf/migrations/0012_add_oauth2_pkce_settings.py:37]: This way doesn't work.
```
Setting.objects.delete(setting="xyz")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'SettingsManager' object has no attribute 'delete'
```
instead I will use list comprehension and `filter(settings__in=[...])`

Comment (rafalp) [reply misago/conf/migrations/0012_add_oauth2_pkce_settings.py:37]: Yes, right 👍

File: misago/oauth2/tests/test_get_code_challenge.py [hunk 22]
```diff
@@ -0,0 +1,28 @@
+import pytest
+
+from .. import exceptions
+from ..client import get_code_challenge
+
+
+def test_exception_is_raised_if_code_verifier_not_provided():
```
Comment (rafalp) [inline misago/oauth2/tests/test_get_code_challenge.py:7]:
```suggestion
def test_exception_is_raised_if_code_verifier_is_not_provided():
```

File: misago/oauth2/tests/test_get_code_challenge.py [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    with pytest.raises(exceptions.OAuth2CodeVerifierNotProvidedError):
+        get_code_challenge(None, "S256")
+
+    with pytest.raises(exceptions.OAuth2CodeVerifierNotProvidedError):
+        get_code_challenge("", "S256")
+
+
+def test_exception_is_raised_if_not_supported_code_challenge_method():
```
Comment (rafalp) [inline misago/oauth2/tests/test_get_code_challenge.py:15]:
```suggestion
def test_exception_is_raised_if_code_challenge_method_is_not_supported():
```

File: misago/oauth2/tests/test_get_code_challenge.py [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+    with pytest.raises(exceptions.OAuth2NotSupportedHashMethodError):
+        get_code_challenge("hYfgN", "ABC")
+
+
+def test_code_challenge_is_returned_S256_hash_method():
```
Comment (rafalp) [inline misago/oauth2/tests/test_get_code_challenge.py:20]:
```suggestion
def test_code_challenge_is_returned_using_S256_hash_method():
```

File: misago/oauth2/tests/test_get_code_challenge.py [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+    code_verifier = "hYfgNABC"
+    code_challenge = "E5AZUNDQx0-aVDeEVgNFTEglcG1bWgOuygG_elkz_io"
+    assert get_code_challenge(code_verifier, "S256") == code_challenge
+
+
+def test_code_challenge_is_returned_plain_hash_method():
```
Comment (rafalp) [inline misago/oauth2/tests/test_get_code_challenge.py:26]:
```suggestion
def test_code_challenge_is_returned_using_plain_hash_method():
```

File: misago/oauth2/client.py [hunk 26]
```diff
@@ -186,3 +203,19 @@ def clear_json_value(value: Any) -> str | None:
         return str(value)
 
     return None
+
+
+def get_code_challenge(code_verifier, code_challenge_method):
```
Comment (rafalp) [inline misago/oauth2/client.py:208]:
```suggestion
def get_code_challenge(code_verifier: str, code_challenge_method: str) -> str:
```

File: misago/oauth2/client.py [hunk 27]
```diff
@@ -186,3 +203,19 @@ def clear_json_value(value: Any) -> str | None:
...
+    if not code_verifier:
```
Comment (rafalp) [inline misago/oauth2/client.py:209]: Is this error even possible?

**Review by rafalp — CHANGES_REQUESTED** [review]

File: misago/conf/migrations/0012_add_oauth2_pkce_settings.py (hunk 2 above)
Comment (raydeal) [reply misago/conf/migrations/0012_add_oauth2_pkce_settings.py:45]: already reformatted

**Review by rafalp — APPROVED** [review]

**Comment by rafalp** [comment]
Thanks!

**Pull request merged**